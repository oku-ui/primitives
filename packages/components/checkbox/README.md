# Checkbox
A control that allows the user to toggle between checked and not checked.

![@oku-ui/checkbox](./../../../.github/assets/og/oku-checkbox.jpg)

[![Version](https://img.shields.io/npm/v/@oku-ui/checkbox?style=flat&colorA=18181B&colorB=28CF8D)](https://www.npmjs.com/package/@oku-ui/checkbox) [![Downloads](https://img.shields.io/npm/dm/@oku-ui/checkbox?style=flat&colorA=18181B&colorB=28CF8D)](https://www.npmjs.com/package/@oku-ui/checkbox)

## Installation

```sh
$ pnpm add @oku-ui/checkbox
```

[Documentation](https://oku-ui.com/primitives/components/checkbox)
