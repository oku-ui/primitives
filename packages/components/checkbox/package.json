{
  "name": "@oku-ui/checkbox",
  "type": "module",
  "version": "0.1.4",
  "license": "MIT",
  "source": "src/index.ts",
  "funding": "https://github.com/sponsors/productdevbook",
  "homepage": "https://oku-ui.com/primitives",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/oku-ui/primitives.git",
    "directory": "packages/components/checkbox"
  },
  "bugs": {
    "url": "https://github.com/oku-ui/primitives/issues"
  },
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs"
    }
  },
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "dev": "tsup --watch"
  },
  "peerDependencies": {
    "vue": "^3.3.4"
  },
  "dependencies": {
    "@oku-ui/primitive": "latest",
    "@oku-ui/provide": "latest",
    "@oku-ui/use-composable": "latest",
    "@oku-ui/utils": "latest"
  },
  "devDependencies": {
    "tsconfig": "workspace:^"
  },
  "engines": {
    "node": ">=18"
  }
}
