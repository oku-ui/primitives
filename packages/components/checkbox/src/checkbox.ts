import { createProvideScope } from '@oku-ui/provide'
import type { PropType, Ref } from 'vue'
import { computed, defineComponent, h, ref, toRefs, useModel, watchEffect } from 'vue'

import { composeEventHandlers } from '@oku-ui/utils'
import { useComposedRefs, useControllable, useForwardRef } from '@oku-ui/use-composable'
import { Primitive, primitiveProps } from '@oku-ui/primitive'

import type { ElementType, PrimitiveProps } from '@oku-ui/primitive'

import type { Scope } from '@oku-ui/provide'
import { getState, isIndeterminate, scopeCheckboxProps } from './utils'
import type { CheckedState, ScopeCheckbox } from './utils'
import { OkuBubbleInput } from './bubbleInput'

const CHECKBOX_NAME = 'OkuCheckbox'

const [createCheckboxProvider, createCheckboxScope] = createProvideScope(CHECKBOX_NAME)

type CheckboxInjectValue = {
  state: Ref<CheckedState>
  disabled?: Ref<boolean | undefined>
}

export const [CheckboxProvider, useCheckboxInject]
  = createCheckboxProvider<CheckboxInjectValue>(CHECKBOX_NAME)

export type CheckboxIntrinsicElement = ElementType<'button'>
export type CheckboxElement = HTMLButtonElement

interface CheckboxProps extends PrimitiveProps {
  checked?: CheckedState
  defaultChecked?: CheckedState
  required?: boolean
  onCheckedChange?(checked: CheckedState): void
  scopeCheckbox?: Scope
  name?: string
  disabled?: boolean
  value?: string
}

const checkboxProps = {
  modelValue: {
    type: [Boolean, String, Number] as PropType<CheckedState>,
    default: undefined,
  },
  checked: {
    type: [Boolean, String, Number] as PropType<CheckedState>,
    default: undefined,
  },
  defaultChecked: {
    type: [Boolean, String] as PropType<boolean | 'indeterminate' | undefined>,
    default: undefined,
  },
  required: {
    type: Boolean as PropType<boolean | undefined>,
    default: undefined,
  },
  name: {
    type: String as PropType<string | undefined>,
    default: undefined,
  },
  disabled: {
    type: Boolean as PropType<boolean | undefined>,
    default: undefined,
  },
  value: {
    type: String as PropType<string | undefined>,
    default: undefined,
  },
}

const checkboxDisplayName = 'OkuCheckbox'

const Checkbox = defineComponent({
  name: checkboxDisplayName,
  components: { OkuBubbleInput },
  inheritAttrs: false,
  props: {
    ...checkboxProps,
    ...scopeCheckboxProps,
    ...primitiveProps,
  },
  emits: ['update:modelValue', 'checkedChange'],
  setup(props, { attrs, slots, emit }) {
    const {
      checked: checkedProp,
      defaultChecked,
      required,
      disabled,
      name,
      value,
    } = toRefs(props)

    const buttonRef = ref<HTMLButtonElement | null>(null)
    const forwardedRef = useForwardRef()
    const composedRefs = useComposedRefs(buttonRef, forwardedRef)

    const {
      ...checkboxProps
    } = attrs as CheckboxIntrinsicElement

    const hasConsumerStoppedPropagationRef = ref(false)

    const modelValue = useModel(props, 'modelValue')
    const proxyChecked = computed({
      get: () => modelValue.value !== undefined ? modelValue.value : checkedProp.value !== undefined ? checkedProp.value : undefined,
      set: () => {
      },
    })

    const { state, updateValue } = useControllable({
      prop: computed(() => proxyChecked.value),
      defaultProp: computed(() => defaultChecked.value),
      onChange: (result: any) => {
        emit('update:modelValue', result)
        emit('checkedChange', result)
      },
    })

    const initialCheckedStateRef = ref(state.value)

    const isFormControl = computed(() => {
      return Boolean(buttonRef.value?.closest('form')) || false
    })

    watchEffect(() => {
      const form = buttonRef.value?.form
      if (form) {
        const reset = () => updateValue(initialCheckedStateRef.value)
        form.addEventListener('reset', reset)
        return () => form.removeEventListener('reset', reset)
      }
    })

    CheckboxProvider({
      scope: props.scopeOkuCheckbox,
      state,
      disabled,
    })

    const originalReturn = () =>
      [h(Primitive.button, {
        'type': 'button',
        'role': 'checkbox',
        'aria-checked': isIndeterminate(state.value) ? 'mixed' : state.value as any,
        'aria-required': required.value,
        'data-state': computed(() => getState(state.value)).value,
        'data-disabled': disabled.value ? '' : undefined,
        'disabled': disabled.value,
        'value': value.value,
        'asChild': props.asChild,
        ...checkboxProps,
        'ref': composedRefs,
        'onKeyDown': composeEventHandlers(checkboxProps.onKeydown, (event) => {
          // According to WAI ARIA, Checkboxes don't activate on enter keypress
          if (event.key === 'Enter')
            event.preventDefault()
        }),
        'onClick': composeEventHandlers(checkboxProps.onClick, (event) => {
          updateValue(isIndeterminate(state.value) ? true : !state.value)

          if (isFormControl.value) {
            // hasConsumerStoppedPropagationRef.value = event.isPropagationStopped()
            // if checkbox is in a form, stop propagation from the button so that we only propagate
            // one click event (from the input). We propagate changes from an input so that native
            // form validation works and form events reflect checkbox updates.
            if (!hasConsumerStoppedPropagationRef.value)
              event.stopPropagation()
          }
        }),
      },
      {
        default: () => slots.default?.(),
      }),
      isFormControl.value && h(
        OkuBubbleInput,
        {
          bubbles: computed(() => !hasConsumerStoppedPropagationRef.value).value,
          name: name.value,
          value: value.value,
          checked: state.value,
          required: required.value,
          control: buttonRef.value,
          disabled: disabled.value,
          // We transform because the input is absolutely positioned but we have
          // rendered it **after** the button. This pulls it back to sit on top
          // of the button.
          style: { transform: 'translateX(-100%)' } as CSSStyleDeclaration,
        },
      ),
      ]

    return originalReturn
  },
})

// TODO: https://github.com/vuejs/core/pull/7444 after delete
export const OkuCheckbox = Checkbox as typeof Checkbox &
(new () => {
  $props: ScopeCheckbox<Partial<CheckboxElement>>
})

export {
  createCheckboxScope,
}

export type {
  CheckboxProps,
}
