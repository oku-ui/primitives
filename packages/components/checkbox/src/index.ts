export {
  OkuCheckbox,
  createCheckboxScope,
  checkboxProps,
} from './checkbox'

export type {
  CheckboxProps,
  CheckboxElement,
  CheckboxIntrinsicElement,
  CheckboxEmits,
} from './checkbox'

export {
  OkuCheckboxIndicator,
  checkboxIndicatorProps,
} from './checkboxIndicator'

export type {
  CheckboxIndicatorElement,
  CheckboxIndicatorProps,
  CheckboxIndicatorIntrinsicElement,
} from './checkboxIndicator'
