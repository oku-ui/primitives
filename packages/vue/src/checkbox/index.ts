export {
  default as OkuCheckbox,
  type CheckboxProps,
  type CheckboxEmits,
} from './Checkbox.vue'
export {
  default as OkuCheckboxIndicator,
  type CheckboxIndicatorProps,
} from './CheckboxIndicator.vue'
