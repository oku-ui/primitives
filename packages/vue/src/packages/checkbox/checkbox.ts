import { computed, defineComponent, h, mergeProps, onMounted, reactive, ref, toRefs, useModel, watchEffect } from 'vue'
import { reactiveOmit, useComposedRefs, useControllable, useForwardRef, useListeners } from '@oku-ui/use-composable'
import { composeEventHandlers } from '@oku-ui/utils'
import { Primitive } from '@oku-ui/primitive'
import { OkuBubbleInput } from './bubble-input'
import { getState, isIndeterminate } from './utils'
import { CHECKBOX_NAME, checkboxProps, checkboxProvider, scopeCheckboxProps } from './props'
import type { CheckboxEmits, CheckboxNativeElement } from './props'

const Checkbox = defineComponent({
  name: CHECKBOX_NAME,
  components: {
    OkuBubbleInput,
  },
  inheritAttrs: false,
  props: {
    ...checkboxProps.props,
    ...scopeCheckboxProps,
  },
  emits: checkboxProps.emits,
  setup(props, { attrs, emit, slots }) {
    const {
      scopeOkuCheckbox,
      name,
      checked: checkedProp,
      defaultChecked,
      required,
      disabled,
      value,
      ...checkboxProps
    } = toRefs(props)

    const _reactive = reactive(checkboxProps)
    const otherProps = reactiveOmit(_reactive, (key, _value) => key === undefined)

    const forwardedRef = useForwardRef()
    const emits = useListeners()

    const buttonRef = ref<HTMLButtonElement | null>(null)
    const composedRefs = useComposedRefs(forwardedRef, buttonRef)
    const hasConsumerStoppedPropagationRef = ref(false)
    // We set this to true by default so that events bubble to forms without JS (SSR)
    const isFormControl = ref<boolean>(false)
    onMounted(() => {
      isFormControl.value = buttonRef.value
        ? typeof buttonRef.value.closest === 'function'
        && Boolean(buttonRef.value.closest('form'))
        : true
    })

    const modelValue = useModel(props, 'modelValue')
    const proxyChecked = computed({
      get: () => modelValue.value !== undefined ? modelValue.value : checkedProp.value !== undefined ? checkedProp.value : undefined,
      set: () => {
      },
    })

    const { state, updateValue } = useControllable({
      prop: computed(() => proxyChecked.value),
      defaultProp: computed(() => defaultChecked.value),
      onChange: (result: any) => {
        modelValue.value = result
        emit('checkedChange', result)
        emit('update:modelValue', result)
      },
      initialValue: false,
    })

    const initialCheckedStateRef = ref(state.value)

    watchEffect((onInvalidate) => {
      const form = buttonRef.value?.form
      if (form) {
        const reset = () => updateValue(initialCheckedStateRef.value)
        form.addEventListener('reset', reset)

        onInvalidate(() => form.removeEventListener('reset', reset))
      }
    })

    checkboxProvider({
      scope: scopeOkuCheckbox.value,
      state,
      disabled,
    })

    return () => [
      h(Primitive.button, {
        'type': 'button',
        'role': 'checkbox',
        'aria-checked': computed(() => isIndeterminate(state.value) ? 'mixed' : state.value).value,
        'aria-required': required.value,
        'data-state': computed(() => getState(state.value)).value,
        'data-disabled': disabled.value ? '' : undefined,
        'disabled': disabled.value,
        'value': value.value,
        ...mergeProps(attrs, otherProps, emits),
        'ref': composedRefs,
        'onKeydown': composeEventHandlers<CheckboxEmits['keydown'][0]>((event) => {
          emit('keydown', event)
        }, (event) => {
          // According to WAI ARIA, Checkboxes don't activate on enter keypress
          if (event.key === 'Enter')
            event.preventDefault()
        }),
        'onClick': composeEventHandlers<CheckboxEmits['click'][0]>((event) => {
          emit('click', event)
        }, (event) => {
          updateValue(isIndeterminate(state.value) ? true : !state.value)

          if (isFormControl.value) {
            // TODO: isPropagationStopped() is not supported in vue
            // hasConsumerStoppedPropagationRef.value = event.isPropagationStopped()

            // if checkbox is in a form, stop propagation from the button so that we only propagate
            // one click event (from the input). We propagate changes from an input so that native
            // form validation works and form events reflect checkbox updates.
            if (!hasConsumerStoppedPropagationRef.value)
              event.stopPropagation()
          }
        }),
      }, () => slots.default?.()),

      isFormControl.value && h(OkuBubbleInput, {
        control: buttonRef.value,
        bubbles: computed(() => !hasConsumerStoppedPropagationRef.value).value,
        name: name.value,
        value: value.value,
        checked: state.value,
        required: required.value,
        disabled: disabled.value,
        // We transform because the input is absolutely positioned but we have
        // rendered it **after** the button. This pulls it back to sit on top
        // of the button.
        style: { transform: 'translateX(-100%)' },
      }),
    ]
  },
})

// TODO: https://github.com/vuejs/core/pull/7444 after delete
export const OkuCheckbox = Checkbox as typeof Checkbox & (new () => { $props: CheckboxNativeElement })
