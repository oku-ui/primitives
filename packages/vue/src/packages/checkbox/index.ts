export { OkuCheckbox } from './checkbox'
export { OkuCheckboxIndicator } from './checkbox-indicator'

export type {
  CheckboxProps,
  CheckboxEmits,
  CheckboxElement,
  CheckboxNativeElement,
} from './props'

export type {
  CheckboxIndicatorProps,
  CheckboxIndicatorElement,
  CheckboxIndicatorNativeElement,
} from './props'
