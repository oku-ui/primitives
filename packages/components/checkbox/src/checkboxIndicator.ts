import type { PropType } from 'vue'
import { computed, defineComponent, h, toRefs } from 'vue'

import { useForwardRef } from '@oku-ui/use-composable'
import { Primitive, primitiveProps } from '@oku-ui/primitive'

import type { ElementType, PrimitiveProps } from '@oku-ui/primitive'

import { OkuPresence } from '@oku-ui/presence'
import type { ScopeCheckbox } from './utils'
import { getState, isIndeterminate, scopeCheckboxProps } from './utils'
import { useCheckboxInject } from './checkbox'

export type CheckboxIndicatorIntrinsicElement = ElementType<'span'>
export type CheckboxIndicatorElement = HTMLSpanElement

export interface CheckboxIndicatorProps extends PrimitiveProps {
  forceMount?: true
}

export const checkboxIndicatorProps = {
  props: {
    forceMount: {
      type: Boolean as PropType<true | undefined>,
      default: undefined,
    },
  },
  emits: {},
}

const INDICATOR_NAME = 'OkuCheckboxIndicator'

const checkboxIndicator = defineComponent({
  name: INDICATOR_NAME,
  props: {
    ...checkboxIndicatorProps.props,
    ...scopeCheckboxProps,
    ...primitiveProps,
  },
  setup(props, { attrs, slots }) {
    const { forceMount } = toRefs(props)
    const { ...indicatorAttrs } = attrs as CheckboxIndicatorIntrinsicElement

    const forwardedRef = useForwardRef()

    const context = useCheckboxInject(INDICATOR_NAME, props.scopeOkuCheckbox)

    const originalReturn = () => h(OkuPresence, {
      present: computed(() => forceMount.value || isIndeterminate(context.state.value) || context.state.value === true).value,
    }, {
      default: () => h(Primitive.span, {
        'ref': forwardedRef,
        'data-state': computed(() => getState(context.state.value)).value,
        'data-disabled': context.disabled?.value ? '' : undefined,
        ...indicatorAttrs,
        'asChild': props.asChild,
        'style': { pointerEvents: 'none', ...attrs.style as any },
      },
      {
        default: () => slots.default?.(),
      }),
    })

    return originalReturn
  },
})

// TODO: https://github.com/vuejs/core/pull/7444 after delete
export const OkuCheckboxIndicator = checkboxIndicator as typeof checkboxIndicator &
(new () => {
  $props: ScopeCheckbox<Partial<CheckboxIndicatorElement>>
})
