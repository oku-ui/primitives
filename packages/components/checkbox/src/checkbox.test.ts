import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import type { Component } from 'vue'
import { h, ref } from 'vue'
import { OkuCheckbox } from './checkbox'
import { OkuCheckboxIndicator } from './checkboxIndicator'
import type { CheckedState } from './utils'

const component = {
  setup(props, { attrs, slots }) {
    return () => h(OkuCheckbox, { ...attrs }, slots)
  },
} as Component

const componentVModel = {
  setup(props, { attrs }) {
    const checked = ref<CheckedState>(false)
    return () => h(OkuCheckbox, {
      ...attrs,
      'modelValue': checked.value,
      'onUpdate:modelValue': e => checked.value = e,
    }, {
      default: () => h(OkuCheckboxIndicator),
    })
  },
} as Component

const componentChecked = {
  setup(props, { attrs }) {
    const checked = ref<CheckedState>(false)
    return () => h(OkuCheckbox, {
      ...attrs,
      checked: checked.value,
      onCheckedChange: e => checked.value = e,
    }, {
      default: () => h(OkuCheckboxIndicator),
    })
  },
} as Component

describe('OkuCheckbox', () => {
  it('renders the component correctly', () => {
    const wrapper = mount(component)
    expect(wrapper.exists()).toBe(true)
  })

  it('renders the component correctly with a label', () => {
    const wrapper = mount(component, {
      slots: {
        default: 'Label',
      },
    })

    expect(wrapper.html()).toContain('<button type="button" role="checkbox" data-state="unchecked">Label</button>')
  })

  it('can be checked', async () => {
    const wrapper = mount(componentVModel)
    expect(wrapper.html()).toContain(`<button type="button" role="checkbox" aria-checked="false" data-state="unchecked">
  <!---->
</button>`)

    const button = wrapper.find('button')
    await button.trigger('click')

    expect(wrapper.html()).toContain('<button type="button" role="checkbox" aria-checked="true" data-state="checked"><span data-state="checked" style="pointer-events: none;"><!----></span></button>')

    await button.trigger('click')

    expect(wrapper.html()).toContain(`<button type="button" role="checkbox" aria-checked="false" data-state="unchecked">
  <!---->
</button>`)
  })

  it('can be checked with checked prop', async () => {
    const wrapper = mount(componentChecked)
    expect(wrapper.html()).toContain(`<button type="button" role="checkbox" aria-checked="false" data-state="unchecked">
  <!---->
</button>`)

    const button = wrapper.find('button')
    await button.trigger('click')

    expect(wrapper.html()).toContain('<button type="button" role="checkbox" aria-checked="true" data-state="checked"><span data-state="checked" style="pointer-events: none;"><!----></span></button>')

    await button.trigger('click')

    expect(wrapper.html()).toContain(`<button type="button" role="checkbox" aria-checked="false" data-state="unchecked">
  <!---->
</button>`)
  })
})
