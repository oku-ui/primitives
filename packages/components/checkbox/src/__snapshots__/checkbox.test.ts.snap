// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OkuCheckbox > can be checked 1`] = `
<div
  data-v-app=""
>
  
  <button
    aria-checked="false"
    data-state="unchecked"
    role="checkbox"
    type="button"
  >
    <!---->
  </button>
  <!---->
  
</div>
`;

exports[`OkuCheckbox > can be checked 2`] = `
<div
  data-v-app=""
>
  
  <button
    aria-checked="true"
    data-state="checked"
    role="checkbox"
    type="button"
  >
    <span
      data-state="checked"
      style="pointer-events: none;"
    >
      <!---->
    </span>
  </button>
  <!---->
  
</div>
`;

exports[`OkuCheckbox > can be checked 3`] = `
<div
  data-v-app=""
>
  
  <button
    aria-checked="false"
    data-state="unchecked"
    role="checkbox"
    type="button"
  >
    <!---->
  </button>
  <!---->
  
</div>
`;

exports[`OkuCheckbox > can be checked with checked prop 1`] = `
<div
  data-v-app=""
>
  
  <button
    aria-checked="false"
    data-state="unchecked"
    role="checkbox"
    type="button"
  >
    <!---->
  </button>
  <!---->
  
</div>
`;

exports[`OkuCheckbox > can be checked with checked prop 2`] = `
<div
  data-v-app=""
>
  
  <button
    aria-checked="true"
    data-state="checked"
    role="checkbox"
    type="button"
  >
    <span
      data-state="checked"
      style="pointer-events: none;"
    >
      <!---->
    </span>
  </button>
  <!---->
  
</div>
`;

exports[`OkuCheckbox > can be checked with checked prop 3`] = `
<div
  data-v-app=""
>
  
  <button
    aria-checked="false"
    data-state="unchecked"
    role="checkbox"
    type="button"
  >
    <!---->
  </button>
  <!---->
  
</div>
`;

exports[`OkuCheckbox > renders the component correctly with a label 1`] = `
<div
  data-v-app=""
>
  
  <button
    aria-checked="false"
    data-state="unchecked"
    role="checkbox"
    type="button"
  >
    Label
  </button>
  <!---->
  
</div>
`;
